File: rock-code/rock.m
%% ROCK-PCA (ROtated Kernel Complex PCA)
%%% for Spatio-temporal data analysis
%%% Diego Bueso, IPL, Universitat de Valencia
%%% diego.bueso@uv.es || (22/05/2019)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [Xpc,pc,expvar,Sigma,Kernel]=rock(f,h,Pk,N)

%% IN PUT
%f --> Spatio temporal data f(Temporal, x Spatial, y Spatial)
%h --> number of PC's [2,3,...,h] 
%N --> number of sigmas to try 
%Pk -> number of powers to try [1,2,...,Pk]


%% OUT PUT
%Xpc    --> Projected ROCK pc
%pc     --> optimus pricipal component from ROCK
%expvar --> Explained variances
%Sigma  --> optimus sigma
%Kernel --> estimated kernel matrix

%% Hilbert transform (complex approach)
[nt,nx,ny]=size(f);
H=eye(nt)-(1/nt)*ones(nt);     % Centering matrix

x=reshape(f,nt,nx*ny);
x=hilbert(x);
x=H*x;                         % Centering data

%% Build Kernel distance matrix across time dimension
k=norm2mat(x',x');
disp('Building kernel matrix');
disp('100 %');

%% Sigma definition; Boundary: median/20 --> maximun·10
m=median(abs(k(:)));    % mean distance
M=max(abs(k(:)));       % maximun distance
sigma=linspace(m/10,M*10,N);

%% Kernel matrix RBF contruction
K=zeros(nt,nt,N);                           % Kernel matrix for each sigma
Pc=zeros(nt,h,N);                           % PC for each sigma
for i=1:N
  K(:,:,i)=exp(-k/(2*sigma(i).^2));         % kernel function              
  K(:,:,i)=H*K(:,:,i)*H;                    % Centering Kernel matrix  
  [v,l]=eig(squeeze(K(:,:,i)));             
  [~,idx]=sort(diag(l),'descend');          
  v=v(:,idx);                               % Sort Pcs                      
  v=v(:,1:h);
  Pc(:,:,i)=v;
  clc;
  disp('Building kernel matrix');
  disp('100 %');
  disp('solving Complex RBF');
  disp([num2str(i*100/N),' %']);
end

%% Kurtosis optimize criterion (number of Components and Sigma)
P=Pk+1;                     % num of power to try
hh=2:h;                     % num of PC to try
Kurtosis=zeros(h-1,N);

for i=hh
for j=1:N
Pcr=Pc(:,1:i,j);
for kk=1:i
     Pcr(:,kk)=Pcr(:,kk)/(Pcr(:,kk)'*Pcr(:,kk));  % Normalize pc's
end
Kurtosis(i,j)=(nt/i)*sum(sum(real(Pcr).^4)./(sum(real(Pcr.^2).^2))); % kurtosis
end
clc;
disp('Building kernel matrix');
disp('100 %');
disp('solving Complex RBF');
disp('100 %');
disp('Kurtosis Optimization');
disp([num2str(i*100/(h+P)),' %']);
end

[~,Npc]=max(max(Kurtosis,[],2));                                % find the number of PC's
[~,Sigma]=max(max(Kurtosis));                                   % find the Sigma

%% Kurtosis optimize criterion (Power of promax rotation)
pc=squeeze(Pc(:,1:Npc,Sigma));
Kurtosis=zeros(P,1);
for p=1:P   
power=p-1;
if power==0
    Pcr=pc;
else
    Pcr=fft(pc);                           % promax rotation on the frequency space 
    Rv=Promax(Pcr,power);
    Rv=(abs(Rv)./abs(Pcr)).*Pcr;
    Rv=hilbert(real(ifft(Rv)));
    Pcr=H*Rv;
end
for kk=1:Npc
     Pcr(:,kk)=Pcr(:,kk)/(Pcr(:,kk)'*Pcr(:,kk));
end
Kurtosis(p)=(nt/i)*sum(sum(real(Pcr).^4)./(sum(real(Pcr.^2).^2)));
clc;
disp('Building kernel matrix');
disp('100 %');
disp('solving Complex RBF');
disp('100 %');
disp('Kurtosis Optimization');
disp([num2str((h+p)*100/(h+P)),' %']);
end

[~,Power]=max(Kurtosis);                        % find the power of promax rotation

disp(['Power: ',num2str(Power)]);
disp(['PCs: ',num2str(Npc)]);
disp(['median(dk)/Sigma: ',num2str(m/sigma(Sigma))]);
disp(['max(dk)/Sigma: ',num2str(M/sigma(Sigma))]);
disp(['sigma: ',num2str(Sigma)]);

%% extract the rotated component
if Power==0
    pc=squeeze(Pc(:,1:Npc,Sigma));
else
    Pcr=squeeze(Pc(:,1:Npc,Sigma));
    Pcr=fft(Pcr);
    Rv=Promax(Pcr,Power);
    Rv=(abs(Rv)./abs(Pcr)).*Pcr;
    Rv=hilbert(real(ifft(Rv)));
    pc=H*Rv;
end

%% reconstruction eigen values
expvar=zeros(Npc,1);
for i=1:Npc
    expvar(i)=norm(pc(:,i)'*K(:,:,Sigma))/norm(pc(:,i));
end
[expvar,idx]=sort(expvar,'descend');
pc=pc(:,idx);

%% spatial projection (covariance approach)
Xpc=pc'*x;
Xpc=reshape(Xpc,Npc,nx,ny);

%% return kernel matrix and his sigma parameter
Kernel=squeeze(K(:,:,Sigma));
Sigma=sigma(Sigma);

end

%% Compute the distance matrix
function D = norm2mat(X1,X2)
    D = - 2 * (X1' * X2);
    D = bsxfun(@plus, D, sum(X1.^2,1)');
    D = bsxfun(@plus, D, sum(X2.^2,1));
end

File: python_code/rock.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

""" Perform Rotated Complex Kernel PCA (ROCK-PCA).

This code is a translation of the original Matlab code into Python.

Original reference:
Bueso, D., Piles, M. & Camps-Valls, G. Nonlinear PCA for Spatio-Temporal
Analysis of Earth Observation Data. IEEE Trans. Geosci. Remote Sensing 1–12
(2020) doi:10.1109/TGRS.2020.2969813.

You can find this code at https://github.com/DiegoBueso/ROCK-PCA.

Author:         Niclas Rieger
Affilation:     Mathematical Research Centre (CRM), Barcelona, Spain
Contact:        nrieger@crm.cat
GitHub:         https://github.com/NiclasRieger

"""
# =============================================================================
# Imports
# =============================================================================
import numpy as np
from scipy.signal import hilbert
from numba import jit, complex128
from tqdm import tqdm
from rotation import promax

# =============================================================================
# Functions
# =============================================================================

@jit(complex128[:,:](complex128[:,:],complex128[:,:]), fastmath=True, parallel=True,boundscheck=False, nopython=True)
def distance_matrix(A,B):
    """
    Calculate distance matrix d_ij = ||a_i-b_i||^2 in space, that is distance
    between stations (columns) over time.

    Parameters
    ----------
    A : array
        First matrix
    B : array
        Second matrix

    Returns
    -------
    D : array
        Distance matrix

    """
    assert A.shape[0] == B.shape[0], 'Distance matrix could not be calculated.'
    ' First dimension of input matrices need to be same.'
    n = A.shape[0]
    p1 = A.shape[1]
    p2 = B.shape[1]
    D = np.zeros((p1, p2),dtype=np.complex128)
    for i in range(p1):
        for j in range(p2):
            s = 0
            for k in range(n):
                s += (A[k, i] - B[k, j])*(A[k, i] - B[k, j]).conjugate()
            D[i, j] = s
    return D


def rock(data, n_rot, n_pow, n_sig):
    """Perform ROCK-PCA.

    Parameters
    ----------
    data : ndarray
        Data with dimensions (temporal, spatial x, spatial y).
    n_rot : int
        Number of PCs to check for optimal rotation.
    n_pow :
        Number of powers to check for optimal rotation.
    n_sig : type
        Number of sigmas to check for optimal kernel matrix estimation.

    Returns
    -------
    projected_data
        Projections of principcal components on data.
    pcs
        Principal components.
    expvar
        Explained variances of PCs.
    sigma
        Optimal sigma.
    kernel
        Estimated kernel matrix.

    """
    nt = data.shape[0]
    nx = data.shape[1]
    ny = data.shape[2]
    ns = np.product((nx, ny))

    data = data.reshape(nt,ns)

    # center data
    center_matrix = np.identity(nt) - (1. / nt) * np.ones(nt)
    data = data - data.mean(axis=0) # computationally more efficient

    # remove NaN in data field (such as continents/ocean)
    no_nan_index = np.where(~(np.isnan(data[0])))[0]
    no_nan_data  = data[:,no_nan_index]

    # complexify data via Hilbert transform
    no_nan_data = hilbert(no_nan_data, axis=0)
    no_nan_data = no_nan_data - no_nan_data.mean(axis=0)


    # Sigma estimation from median distance
    print('Build distance matrix ... ', flush=True)
    distance = distance_matrix(no_nan_data.T, no_nan_data.T)
    median_distance = np.median(abs(distance))
    max_distance    = np.max(abs(distance))
    sigmas          = np.linspace(.1 * median_distance, 10 * max_distance,n_sig)

    # Kernel matrix RBF
    kernels = np.zeros((n_sig, nt, nt), dtype=complex)
    pcs     = np.zeros((n_sig, nt, n_rot), dtype=complex)

    description = 'Build kernel matrices'
    for s, sigma in tqdm(enumerate(sigmas), total=len(sigmas), desc=description):
        K = np.exp( -distance / (2 * (sigma**2)) )
        K = center_matrix @ K @ center_matrix
        U, eigenvalues, _ = np.linalg.svd(K, full_matrices=False)

        kernels[s,:,:] = K
        pcs[s,:,:] = U[:,:n_rot]

    # Kurtosis optimization criteria
    powers      = range(0,n_pow+1) # power = 0 means no rotation
    rotations   = range(2,n_rot+1)

    kurtosis = np.zeros((len(sigmas),len(rotations)))

    # optimize sigma & number of PCs
    description = 'Optimize SIGMA and ROTATION'
    for s,sig in tqdm(enumerate(sigmas), total=len(sigmas), desc=description):
        for r,rot in enumerate(rotations):
            temp_pcs = pcs[s,:,:(rot)]
            temp_pcs = temp_pcs / np.linalg.norm(temp_pcs, axis=0)**2
            kurtosis[s,r] = (nt/rot)*np.sum(np.sum(np.real(temp_pcs)**4,axis=0)/(np.sum(np.real(temp_pcs**2),axis=0)**2))

    # find maximum kurtosis
    idx_optimal_rot     = np.argmax(np.max(kurtosis, axis=0))
    idx_optimal_sig   = np.argmax(np.max(kurtosis, axis=1))

    optimal_rot     = rotations[idx_optimal_rot]
    optimal_sig     = sigmas[idx_optimal_sig]
    optimal_kernel  = kernels[idx_optimal_sig,:,:]

    # optimize power of Promax
    power_pcs = np.zeros((len(powers),nt,optimal_rot), dtype=complex)
    kurtosis = np.zeros((len(powers)))

    description = 'Optimize POWER'

    for i,pow in tqdm(enumerate(powers), total=len(powers), desc=description):

        temp_pcs = pcs[idx_optimal_sig,:,:optimal_rot]

        if ( pow > 0):
            fft_pcs         = np.fft.fft(temp_pcs, axis=0)
            temp_pcs,_,_    = promax(fft_pcs, power=pow)
            temp_pcs        = abs(temp_pcs) / abs(fft_pcs) * fft_pcs
            temp_pcs        = np.real(np.fft.ifft(temp_pcs, axis=0))
            temp_pcs        = hilbert(temp_pcs, axis=0)
            temp_pcs        = center_matrix @ temp_pcs

        power_pcs[i,:,:] = temp_pcs

        temp_pcs = temp_pcs / np.linalg.norm(temp_pcs,axis=0)**2
        kurtosis[i] = (nt/optimal_rot)*np.sum(np.sum(np.real(temp_pcs)**4,axis=0)/(np.sum(np.real(temp_pcs**2),axis=0)**2))


    # find maximum kurtosis
    idx_optimal_pow = np.argmax(kurtosis)
    optimal_pow = powers[idx_optimal_pow]

    # select PCs accordingly to optimized power
    pcs = power_pcs[idx_optimal_pow,:,:]

    # reconstuction Eigenvalues
    expvar = np.zeros(optimal_rot)
    for i, pc in enumerate(pcs.T):
        expvar[i] = np.linalg.norm(pc.conjugate().T @ optimal_kernel) \
        / np.linalg.norm(pc)

    idx_eig = np.argsort(expvar)
    expvar  = expvar[idx_eig][::-1]
    pcs     = pcs[:,idx_eig][:,::-1]

    # recreate original data dimensions with nan values (continents, oceans)
    data_proj = np.zeros((optimal_rot,ns), dtype=complex) * np.nan

    # spatial projection
    data_proj[:,no_nan_index] = (pcs.conjugate().T @ no_nan_data)
    data_proj = data_proj.reshape(optimal_rot, nx, ny)

    print(optimal_sig)
    print(optimal_rot)
    print(optimal_pow)
    print('Finished')

    return data_proj.real, pcs, expvar, optimal_sig, optimal_kernel
